File: egg-niuke-training/16-longest-palindromic-substring.py
"""
NC17 最长回文子串
https://www.nowcoder.com/practice/b4525d1d84934cf280439aeecc36f4af

对于长度为n的一个字符串A，找到一个最长的回文子串。

解法1：中心扩展法 O(n^2)
解法2：动态规划 O(n^2)
解法3：Manacher算法 O(n)

时间复杂度：O(n^2) - 中心扩展
空间复杂度：O(1)
"""

def longest_palindrome_center(s):
    """
    方法1：中心扩展法
    """
    if not s:
        return ""
    
    def expand_around_center(left, right):
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        return left + 1, right - 1
    
    start = end = 0
    
    for i in range(len(s)):
        # 奇数长度回文
        left1, right1 = expand_around_center(i, i)
        # 偶数长度回文
        left2, right2 = expand_around_center(i, i + 1)
        
        # 选择更长的
        if right1 - left1 > end - start:
            start, end = left1, right1
        if right2 - left2 > end - start:
            start, end = left2, right2
    
    return s[start:end+1]

def longest_palindrome_dp(s):
    """
    方法2：动态规划
    dp[i][j]表示s[i:j+1]是否为回文
    """
    if not s:
        return ""
    
    n = len(s)
    dp = [[False] * n for _ in range(n)]
    start, max_len = 0, 1
    
    # 单个字符都是回文
    for i in range(n):
        dp[i][i] = True
    
    # 枚举子串长度
    for length in range(2, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1
            
            if s[i] == s[j]:
                if length == 2:
                    dp[i][j] = True
                else:
                    dp[i][j] = dp[i+1][j-1]
            
            if dp[i][j] and length > max_len:
                start = i
                max_len = length
    
    return s[start:start+max_len]

# 测试
if __name__ == "__main__":
    test_cases = [
        "babad",
        "cbbd",
        "a",
        "ac"
    ]
    
    for s in test_cases:
        print(f"字符串: {s}")
        print(f"中心扩展: {longest_palindrome_center(s)}")
        print(f"动态规划: {longest_palindrome_dp(s)}\n")

File: egg-niuke-training/16-longest-palindromic-substring.cpp
/**
 * NC17 最长回文子串
 * https://www.nowcoder.com/practice/b4525d1d84934cf280439aeecc36f4af
 * 
 * 对于长度为n的一个字符串A，找到一个最长的回文子串。
 * 
 * 时间复杂度：O(n^2)
 * 空间复杂度：O(1)
 */

#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    /**
     * 方法1：中心扩展法
     */
    string longestPalindrome_Center(string s) {
        if (s.empty()) return "";
        
        int start = 0, maxLen = 1;
        
        for (int i = 0; i < s.length(); i++) {
            // 奇数长度回文
            auto [left1, right1] = expandAroundCenter(s, i, i);
            // 偶数长度回文
            auto [left2, right2] = expandAroundCenter(s, i, i + 1);
            
            // 选择更长的
            if (right1 - left1 + 1 > maxLen) {
                start = left1;
                maxLen = right1 - left1 + 1;
            }
            if (right2 - left2 + 1 > maxLen) {
                start = left2;
                maxLen = right2 - left2 + 1;
            }
        }
        
        return s.substr(start, maxLen);
    }
    
    /**
     * 方法2：动态规划
     */
    string longestPalindrome_DP(string s) {
        if (s.empty()) return "";
        
        int n = s.length();
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        int start = 0, maxLen = 1;
        
        // 单个字符都是回文
        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
        }
        
        // 枚举子串长度
        for (int len = 2; len <= n; len++) {
            for (int i = 0; i <= n - len; i++) {
                int j = i + len - 1;
                
                if (s[i] == s[j]) {
                    if (len == 2) {
                        dp[i][j] = true;
                    } else {
                        dp[i][j] = dp[i+1][j-1];
                    }
                }
                
                if (dp[i][j] && len > maxLen) {
                    start = i;
                    maxLen = len;
                }
            }
        }
        
        return s.substr(start, maxLen);
    }

private:
    pair<int, int> expandAroundCenter(const string& s, int left, int right) {
        while (left >= 0 && right < s.length() && s[left] == s[right]) {
            left--;
            right++;
        }
        return {left + 1, right - 1};
    }
};

int main() {
    Solution solution;
    
    vector<string> testCases = {"babad", "cbbd", "a", "ac"};
    
    for (const string& s : testCases) {
        cout << "字符串: " << s << endl;
        cout << "中心扩展: " << solution.longestPalindrome_Center(s) << endl;
        cout << "动态规划: " << solution.longestPalindrome_DP(s) << endl << endl;
    }
    
    return 0;
}

